runners/java-job-service/src/main/java/org/apache/beam/runners/jobsubmission/InMemoryJobService.java

```diff
@@ -287,26 +283,50 @@ public void run(RunJobRequest request, StreamObserver<RunJobResponse> responseOb
     }
   }
 
+  private Map<String, List<RunnerApi.ArtifactInformation>> extractDependencies(
+      RunnerApi.Pipeline pipeline) {
+    Map<String, List<RunnerApi.ArtifactInformation>> dependencies = new HashMap<>();
+    for (Map.Entry<String, RunnerApi.Environment> entry :
+        pipeline.getComponents().getEnvironmentsMap().entrySet()) {
+      List<RunnerApi.Environment> subEnvs = Environments.expandAnyOfEnvironments(entry.getValue());
+      for (int i = 0; i < subEnvs.size(); i++) {
+        dependencies.put(i + ":" + entry.getKey(), subEnvs.get(i).getDependenciesList());
+      }
+    }
+    return dependencies;
+  }
+
   private RunnerApi.Pipeline resolveDependencies(RunnerApi.Pipeline pipeline, String stagingToken) {
     Map<String, List<RunnerApi.ArtifactInformation>> resolvedDependencies =
```

Comment:
Instead of allowing null here can we default to an empty map if `stagingService.getService().getStagedArtifacts(stagingToken)` returns `null`?

or wrap in an optional

can mark as `@Nullable` if the other 2 are not feasible

Reply:
`null` seems to indicate that no staging was attempted here (which is technically fine for something that has no artifacts, but we can't assume that if there's no entry). This is different from "the dependencies resolved to the empty set" which would be an error to just accept. 

Marking it as @Nullable is painful as the null checker isn't smart enough to figure out that it'll never be null in the second clause.

Reply:
Ack thanks!